### PR TITLE
[Uniques V2] Smart attributes

### PR DESCRIPTION
**Pull request by jsidorenko**
This PR adds a new feature called Smart Attributes.
The main idea of that feature is to allow an owner of an NFT to allow other entities (e.g. another account, an application, a parachain, a custom origin, etc.) to update certain attributes about the NFT. For example, a member of the Polkadot Technical Fellowship could have a membership badge that evolves over time by giving the Fellowship origin permission to update its attributes (e.g. Rank) on the badge directly.

On a technical level to achieve all that I decided to prefix attributes with the namespace.
Currently, 4 different namespaces were created:
- `Pallet(PalletId)` - which means the attribute was set and could be changed by the pallet only;
- `CollectionOwner` - means the attribute can be set and modified by the collection's owner only;
- `ItemOwner` - a namespace for the NFT owners. This is an ideal place to store some custom values like dapps preferences, dids, NFT's custom title, etc.;
- `Account(AccountId)` - attributes within that namespace could be modified only if the `origin` was given permission to do so.

To control whether an external account could modify attributes within its namespace the new methods were added:
`approve_item_attributes` and `cancel_item_attributes_approval`. When the permission gets revoked, all the previously set attributes by that account will be removed.

### CODE REVIEW

File: frame/nfts/Cargo.toml [hunk 1]
```diff
@@ -20,14 +20,14 @@ scale-info = { version = "2.1.1", default-features = false, features = ["derive"
 frame-benchmarking = { version = "4.0.0-dev", default-features = false, optional = true, path = "../benchmarking" }
 frame-support = { version = "4.0.0-dev", default-features = false, path = "../support" }
 frame-system = { version = "4.0.0-dev", default-features = false, path = "../system" }
-sp-runtime = { version = "6.0.0", default-features = false, path = "../../primitives/runtime" }
-sp-std = { version = "4.0.0", default-features = false, path = "../../primitives/std" }
+sp-runtime = { version = "7.0.0", default-features = false, path = "../../primitives/runtime" }
+sp-std = { version = "5.0.0", default-features = false, path = "../../primitives/std" }
```
Comment (gilescope) [inline frame/nfts/Cargo.toml:24]: we should either upgrade it for all of substrate or none of substrate. And if we are upgrading it, I'd suggest doing that in a separate PR.

File: frame/support/src/traits/tokens/nonfungible_v2.rs [hunk 2]
```diff
@@ -42,15 +45,23 @@ pub trait Inspect<AccountId> {
 	/// Returns the attribute value of `item` corresponding to `key`.
 	///
 	/// By default this is `None`; no attributes are defined.
-	fn attribute(_item: &Self::ItemId, _key: &[u8]) -> Option<Vec<u8>> {
+	fn attribute(
+		_item: &Self::ItemId,
+		_key: &[u8],
+		_namespace: &AttributeNamespace<AccountId>,
+	) -> Option<Vec<u8>> {
```
Comment (gilescope) [inline frame/support/src/traits/tokens/nonfungible_v2.rs:52]: should arg order logically go item, namespace, key? as there's many keys in a namespace

Comment (jsidorenko) [reply frame/support/src/traits/tokens/nonfungible_v2.rs:52]: good point, changing that

File: frame/nfts/src/lib.rs [hunk 3]
```diff
@@ -1230,11 +1265,41 @@ pub mod pallet {
...
+		}
+
+		/// Force-set an attribute for a collection or item.
+		///
+		/// Origin must be `ForceOrigin`.
+		///
+		/// If the attribute already exist and it was set by another account, the deposit
+		/// will be returned to the previous owner.
+		///
+		/// - `set_as`: An optional owner if the attribute.
```
Comment (gilescope) [inline frame/nfts/src/lib.rs:1281]:
```suggestion
		/// - `set_as`: An optional owner of the attribute.
```

File: frame/support/src/traits/tokens/nonfungible_v2.rs (hunk 2 above)
Comment (jsidorenko) [reply frame/support/src/traits/tokens/nonfungible_v2.rs:52]: done

File: frame/nfts/src/lib.rs [hunk 4]
```diff
@@ -1230,11 +1265,41 @@ pub mod pallet {
...
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::force_set_attribute())]
+		pub fn force_set_attribute(
+			origin: OriginFor<T>,
+			set_as: Option<T::AccountId>,
+			collection: T::CollectionId,
+			maybe_item: Option<T::ItemId>,
+			key: BoundedVec<u8, T::KeyLimit>,
+			value: BoundedVec<u8, T::ValueLimit>,
+			namespace: AttributeNamespace<T::AccountId>,
```
Comment (gilescope) [inline frame/nfts/src/lib.rs]: again here logically it should go collection, maybe_item, namespace, key, value

File: frame/nfts/src/lib.rs [hunk 5]
```diff
@@ -1257,11 +1323,55 @@ pub mod pallet {
...
+
+		/// Approve item's attributes to be changed by a delegated third-party account.
+		///
+		/// Origin must be Signed and must be an owner of the `item`.
+		///
+		/// - `collection`: A collection of the item.
+		/// - `item`: The item that holds attributes.
+		/// - `delegate`: The account to delegate permission to change attributes of the item.
+		///
+		/// Emits `ItemAttributesApprovalAdded` on success.
```
Comment (gilescope) [inline frame/nfts/src/lib.rs:1342]: can approval of an item be signed by the owner for later use on chain by the delegated third-party? If so would be great to mention this in the docs here.

File: frame/nfts/src/lib.rs [hunk 6]
```diff
@@ -1257,11 +1323,55 @@ pub mod pallet {
...
+			let origin = ensure_signed(origin)?;
+			let delegate = T::Lookup::lookup(delegate)?;
+			Self::do_approve_item_attributes(origin, collection, item, delegate)
+		}
+
+		/// Cancel the previously provided approval to change item's attributes.
+		///
+		/// Origin must be Signed and must be an owner of the `item`.
+		///
+		/// - `collection`: A collection of the item.
```
Comment (gilescope) [inline frame/nfts/src/lib.rs:1359]:
```suggestion
		/// - `collection`: Collection that the item is contained within.
```

File: frame/nfts/src/features/attributes.rs [hunk 7]
```diff
@@ -34,90 +35,273 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+			T::Currency::unreserve(&origin, old_deposit.amount - deposit);
 		}
 
-		Attribute::<T, I>::insert((&collection, maybe_item, &key), (&value, deposit));
+		// NOTE: we don't track the depositor in the CollectionOwner namespace as it's always a
+		// collection's owner. This simplifies the collection's transfer to another owner.
+		let deposit_owner = match namespace {
+			AttributeNamespace::CollectionOwner => {
+				collection_details.owner_deposit.saturating_reduce(old_deposit.amount);
+				collection_details.owner_deposit.saturating_accrue(deposit);
```
Comment (gilescope) [inline frame/nfts/src/features/attributes.rs:100]: wondering if these two lines would be slightly safer switched around so that the accrue happened first?

Comment (jsidorenko) [reply frame/nfts/src/features/attributes.rs:99]: changed

File: frame/nfts/src/features/attributes.rs [hunk 8]
```diff
@@ -34,90 +35,273 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+				},
+			AttributeNamespace::Account(account_id) =>
+				if let Some(item) = maybe_item {
+					let approvals = ItemAttributesApprovalsOf::<T, I>::get(&collection, &item);
+					result = account_id == origin && approvals.contains(&origin)
+				},
+			_ => (),
+		};
+		Ok(result)
+	}
```
[the remark this answers is not in the document]
Comment (tonyalaribe) [reply frame/nfts/src/features/attributes.rs:306]: Actually, now I think of it, the approach I suggested would likely need some more work with the `if let Some()` expressions, so it's probably better to ignore the suggestion.

[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/nfts/src/features/attributes.rs:306]: aha, I first removed the result and had a lot of `} else {` block those return false. Then I added the `result` and the code became clearer

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2058403

File: frame/nfts/src/features/attributes.rs [hunk 9]
```diff
@@ -34,90 +35,273 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+				},
+				Some(item) => {
+					let maybe_is_locked = Self::get_item_config(&collection, &item)
+						.map(|c| c.has_disabled_setting(ItemSetting::UnlockedAttributes))?;
+					ensure!(!maybe_is_locked, Error::<T, I>::LockedItemAttributes);
+				},
 			},
-		};
+			_ => (),
+		}
```
Comment (tonyalaribe) [inline frame/nfts/src/features/attributes.rs:66]: I'm slightly surprised that this works without a semicolon after the closing `match` brace. 
But i'm fairly new to rust. My expectation was that not adding a semicolon here would mean an early return.

Comment (jsidorenko) [reply frame/nfts/src/features/attributes.rs:66]: You need a semicolon when you use `match` as an expression: let smth = match {};

**Comment by jsidorenko** [comment]
bot merge

**Pull request merged**